File: prep/implementation/iterations/iteration_cpp/binary_tree_implicit_array.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <queue>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

template <typename T> struct Node {
  T data;
  Node(T data) { this->data = data; }
};

template <typename T> class BinaryTree {
private:
  vector<Node<T> *> arr;

  void build(int node, const vector<string> &nodes,
             function<T(string)> parser) {
    if (node >= nodes.size() || nodes[node] == "null") {
      arr[node] = nullptr;
      return;
    }

    if (arr[node] == nullptr) {
      arr[node] = new Node(parser(nodes[node]));
    }

    if (2 * node + 1 < nodes.size()) {
      build(2 * node + 1, nodes, parser);
    }

    if (2 * node + 2 < nodes.size()) {
      build(2 * node + 2, nodes, parser);
    }
  }

  int height(int node) {
    if (arr[node] == nullptr) {
      return 0;
    }

    return 1 + max(height(2 * node + 1), height(2 * node + 2));
  }

public:
  BinaryTree(const vector<string> &nodes, function<T(string)> parser) {
    for (int i = 0; i < nodes.size(); i++) {
      this->arr.push_back(nullptr);
    }

    build(0, nodes, parser);
  }

  ~BinaryTree() {
    for (auto node_ptr : this->arr) {
      if (!node_ptr) {
        // delete node_ptr;
      }
    }
  }

  void dfs_pre(int node) {
    if (arr[node] == nullptr) {
      return;
    }

    cout << arr[node]->data << " ";

    if (2 * node + 1 < arr.size()) {
      dfs_pre(2 * node + 1);
    }

    if (2 * node + 2 < arr.size()) {
      dfs_pre(2 * node + 2);
    }
  }

  void dfs_pre() {
    dfs_pre(0);
    cout << "\n";
  }

  void dfs_pre_iter() {
    if (arr[0] == nullptr) {
      return;
    }

    vector<int> st;
    st.push_back(0);

    while (!st.empty()) {
      int node = st.back();
      st.pop_back();

      cout << arr[node]->data << " ";

      if (2 * node + 2 < arr.size() && arr[2 * node + 2] != nullptr) {
        st.push_back(2 * node + 2);
      }

      if (2 * node + 1 < arr.size() && arr[2 * node + 1] != nullptr) {
        st.push_back(2 * node + 1);
      }
    }
    cout << "\n";
  }

  void dfs_in(int node) {
    if (arr[node] == nullptr) {
      return;
    }

    if (2 * node + 1 < arr.size() && arr[2 * node + 1] != nullptr) {
      dfs_in(2 * node + 1);
    }

    cout << arr[node]->data << " ";

    if (2 * node + 2 < arr.size() && arr[2 * node + 2] != nullptr) {
      dfs_in(2 * node + 2);
    }
  }

  void dfs_in() {
    dfs_in(0);
    cout << "\n";
  }

  void dfs_in_iter() {
    vector<int> st;
    int curr_node = 0;

    while ((curr_node < arr.size() && arr[curr_node] != nullptr) ||
           !st.empty()) {
      while (curr_node < arr.size() && arr[curr_node] != nullptr) {
        st.push_back(curr_node);
        curr_node = 2 * curr_node + 1;
      }

      if (!st.empty()) {
        curr_node = st.back();
        st.pop_back();
        cout << arr[curr_node]->data << " ";
        curr_node = 2 * curr_node + 2;
      }
    }
    cout << "\n";
  }

  void dfs_post(int node) {
    if (arr[node] == nullptr) {
      return;
    }

    if (2 * node + 1 < arr.size()) {
      dfs_post(2 * node + 1);
    }

    if (2 * node + 2 < arr.size()) {
      dfs_post(2 * node + 2);
    }

    cout << arr[node]->data << " ";
  }

  void dfs_post() {
    dfs_post(0);
    cout << "\n";
  }

  void dfs_post_iter() {
    vector<int> st;
    vector<int> rev_st;

    st.push_back(0);
    while (!st.empty()) {
      int node = st.back();
      st.pop_back();

      rev_st.push_back(node);

      if (2 * node + 1 < arr.size() && arr[2 * node + 1] != nullptr) {
        st.push_back(2 * node + 1);
      }

      if (2 * node + 2 < arr.size() && arr[2 * node + 2] != nullptr) {
        st.push_back(2 * node + 2);
      }
    }

    while (!rev_st.empty()) {
      cout << arr[rev_st.back()]->data << " ";
      rev_st.pop_back();
    }
    cout << "\n";
  }

  void bfs() {
    queue<int> q;
    q.push(0);

    while (!q.empty()) {
      int node = q.front();
      q.pop();

      cout << arr[node]->data << " ";

      if (2 * node + 1 < arr.size() && arr[2 * node + 1] != nullptr) {
        q.push(2 * node + 1);
      }

      if (2 * node + 2 < arr.size() && arr[2 * node + 2] != nullptr) {
        q.push(2 * node + 2);
      }
    }
    cout << "\n";
  }

  int height() { return height(0); }

  void bfs_rec(int node, int level) {
    if (arr[node] == nullptr) {
      return;
    }

    if (level == 0) {
      cout << arr[node]->data << " ";
    } else {
      if (2 * node + 1 < arr.size()) {
        bfs_rec(2 * node + 1, level - 1);
      }

      if (2 * node + 2 < arr.size()) {
        bfs_rec(2 * node + 2, level - 1);
      }
    }
  }

  void bfs_rec() {
    int h = height();
    for (int i = 0; i < h; i++) {
      bfs_rec(0, i);
    }
    cout << "\n";
  }
};

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  string input;
  getline(cin, input);

  stringstream ss(input);
  vector<string> nodes;
  string tmp;

  while (getline(ss, tmp, ' ')) {
    nodes.push_back(tmp);
  }

  BinaryTree<int> tree(nodes, [](const string &s) { return stoi(s); });

  tree.dfs_pre();
  tree.dfs_pre_iter();

  tree.dfs_in();
  tree.dfs_in_iter();

  tree.dfs_post();
  tree.dfs_post_iter();

  tree.bfs();
  tree.bfs_rec();

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/main_1/BinaryTreeImplicitArray.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class BinaryTreeImplicitArray {
  static class BinaryTree {
    static class TreeNode {
      int val;

      TreeNode(int val) { this.val = val; }
    }

    private List<TreeNode> arr;
    private int size;

    public BinaryTree(String[] nodes) {
      this.size = nodes.length;

      this.arr = new ArrayList<>(this.size);
      for (int i = 0; i < size; i++) {
        this.arr.add(null);
      }

      buildTree(nodes, 0);
    }

    private void buildTree(String[] nodes, int currIdx) {
      if (currIdx >= this.size || nodes[currIdx].equalsIgnoreCase("null")) {
        return;
      } else {
        this.arr.set(currIdx, new TreeNode(Integer.parseInt(nodes[currIdx])));
        buildTree(nodes, 2 * currIdx + 1);
        buildTree(nodes, 2 * currIdx + 2);
      }
    }

    // O(n) / O(n)
    public void dfsPre() {
      if (this.arr.get(0) == null) {
        return;
      }

      dfsPre(0);
      System.out.println();
    }

    private void dfsPre(int node) {
      if (node >= this.size || this.arr.get(node) == null) {
        return;
      }

      System.out.printf("%d ", this.arr.get(node).val);
      dfsPre(2 * node + 1);
      dfsPre(2 * node + 2);
    }

    // O(n) / O(n)
    public void dfsPreIter() {
      Stack<Integer> st = new Stack<>();
      st.push(0);

      while (!st.isEmpty()) {
        int node = st.pop();

        System.out.printf("%d ", this.arr.get(node).val);

        if (2 * node + 2 < this.size && this.arr.get(2 * node + 2) != null) {
          st.push(2 * node + 2);
        }

        if (2 * node + 1 < this.size && this.arr.get(2 * node + 1) != null) {
          st.push(2 * node + 1);
        }
      }

      System.out.println();
    }

    // O(n) / O(n)
    public void dfsIn() {
      if (this.arr.get(0) == null) {
        return;
      }

      dfsIn(0);
      System.out.println();
    }

    private void dfsIn(int node) {
      if (node >= this.size || this.arr.get(node) == null) {
        return;
      }

      dfsIn(2 * node + 1);
      System.out.printf("%d ", this.arr.get(node).val);
      dfsIn(2 * node + 2);
    }

    // O(n) / O(n)
    public void dfsInIter() {
      if (this.arr.get(0) == null) {
        return;
      }

      Stack<Integer> st = new Stack<>();
      int currNode = 0;

      while ((currNode < this.size && this.arr.get(currNode) != null) ||
             !st.isEmpty()) {
        while (currNode < this.size && this.arr.get(currNode) != null) {
          st.push(currNode);
          currNode = 2 * currNode + 1;
        }

        if (!st.isEmpty()) {
          currNode = st.pop();
          System.out.printf("%d ", this.arr.get(currNode).val);
          currNode = 2 * currNode + 2;
        }
      }
      System.out.println();
    }

    // O(n) / O(n)
    public void dfsPost() {
      if (this.arr.get(0) == null) {
        return;
      }

      dfsPost(0);
      System.out.println();
    }

    private void dfsPost(int node) {
      if (node >= this.size || this.arr.get(node) == null) {
        return;
      }

      dfsPost(2 * node + 1);
      dfsPost(2 * node + 2);
      System.out.printf("%d ", this.arr.get(node).val);
    }

    public void dfsPostIter() {
      if (this.arr.get(0) == null) {
        return;
      }

      Stack<Integer> st = new Stack<>();
      Stack<Integer> revSt = new Stack<>();

      st.push(0);

      while (!st.isEmpty()) {
        int node = st.pop();
        revSt.push(node);

        if (2 * node + 1 < this.size && this.arr.get(2 * node + 1) != null) {
          st.push(2 * node + 1);
        }

        if (2 * node + 2 < this.size && this.arr.get(2 * node + 2) != null) {
          st.push(2 * node + 2);
        }
      }

      while (!revSt.isEmpty()) {
        System.out.printf("%d ", this.arr.get(revSt.pop()).val);
      }
      System.out.println();
    }

    // O(n) / O(n)
    public void bfs() {
      if (this.arr.get(0) == null) {
        return;
      }

      Queue<Integer> q = new ArrayDeque<>();
      q.add(0);

      while (!q.isEmpty()) {
        int node = q.poll();

        System.out.printf("%d ", this.arr.get(node).val);

        if (2 * node + 1 < this.size && this.arr.get(2 * node + 1) != null) {
          q.add(2 * node + 1);
        }

        if (2 * node + 2 < this.size && this.arr.get(2 * node + 2) != null) {
          q.add(2 * node + 2);
        }
      }

      System.out.println();
    }

    // O(n) / O(n)
    public void bfsRec() {
      if (this.arr.get(0) == null) {
        return;
      }

      Queue<Integer> q = new ArrayDeque<>();
      q.add(0);

      bfsRec(q);
      System.out.println();
    }

    private void bfsRec(Queue<Integer> q) {
      if (q.isEmpty()) {
        return;
      }

      int node = q.poll();
      System.out.printf("%d ", this.arr.get(node).val);

      if (2 * node + 1 < this.size && this.arr.get(2 * node + 1) != null) {
        q.add(2 * node + 1);
      }

      if (2 * node + 2 < this.size && this.arr.get(2 * node + 2) != null) {
        q.add(2 * node + 2);
      }

      bfsRec(q);
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      String[] nodes = in.nextLine().trim().split(" ");

      BinaryTree tree = new BinaryTree(nodes);
      tree.dfsPre();
      tree.dfsPreIter();

      tree.dfsIn();
      tree.dfsInIter();

      tree.dfsPost();
      tree.dfsPostIter();

      tree.bfs();
      tree.bfsRec();
    }
  }
}